xds/internal/xdsclient/watchers_test.go

```diff
@@ -0,0 +1,599 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xdsclient
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/xds/internal/xdsclient/bootstrap"
+	"google.golang.org/grpc/xds/internal/xdsclient/pubsub"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource/version"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+// testWatchSetup starts the client and makes a watch on it for the resourceName.
+// The caller can make new watchers, or start sending updates for this watch
+// afterwards.
+//
+// Note that this function checks for the controller after watching. This is the
+// new behavior after federation.
+func testWatchSetup(ctx context.Context, t *testing.T, typ xdsresource.ResourceType, resourceName string, overrideWatchExpiryTimeout bool) (_ *clientImpl, _ *testController, _ pubsub.UpdateHandler, updateCh *testutils.Channel, cancelWatch func()) {
+	t.Helper()
+
+	apiClientCh, cleanupController := overrideNewController()
+	t.Cleanup(cleanupController)
+
+	watchExpiryTimeout := defaultWatchExpiryTimeout
+	if overrideWatchExpiryTimeout {
+		watchExpiryTimeout = defaultTestWatchExpiryTimeout
+	}
+
+	client, err := newWithConfig(clientOpts(), watchExpiryTimeout, defaultIdleAuthorityDeleteTimeout)
+	if err != nil {
+		t.Fatalf("failed to create client: %v", err)
+	}
+	t.Cleanup(client.Close)
+
+	newWatchF, _, _ := typeToTestFuncs(typ)
+	updateCh, cancelWatch = newWatchF(client, resourceName)
+	t.Cleanup(cancelWatch)
```

Comment:
Why do we have to return this func if `t.Cleanup` is already called on it?

Reply:
Returned func is called by tests (for behaviors after watch is canceled).